### DISCUSSION TITLE
Some tests fail with some supported httpx versions due to whitespace differences in JSON responses

### DISCUSSION THREAD

**Discussion by musicinmybrain** [encode/starlette#2795] [Potential Issue] [answered]
I’m working on updating the [`python-starlette` package](https://src.fedoraproject.org/rpms/python-starlette) to 0.42.0 in Fedora.

I see that https://github.com/encode/starlette/pull/2773 pinned `httpx` to the range `>=0.27.0,<0.29.0` due what seems to have been [an API incompatibility](https://github.com/encode/starlette/issues/2771).

That PR also adjusted the whitespace in some expected JSON response strings, e.g. from `{"body": '{"a": "123"}'}` to `{"body": '{"a":"123"}'}` This seems to be correct for the latest `httpx`, 0.28.1, but the whitespace change isn’t in the `httpx` 0.27 series, so several tests fail if `httpx` 0.27.0, 0.27.1, or 0.27.2 is installed. For example:

```
$ git clone https://github.com/encode/starlette.git
$ cd starlette
$ git checkout 0.42.0
$ uv venv _e
$ . _e/bin/activate
(_e) $ uv pip install -e .[full]
(_e) $ uv pip install -r requirements.txt
(_e) $ python -m pytest
= 867 passed, 2 xfailed in 4.18s =
(_e) $ uv pip install httpx==0.27.2
(_e) $ python -m pytest
[output condensed with most details omitted]
= FAILURES =
_ test_wsgi_post[asyncio] _
>       assert response.json() == {"body": '{"a":"123"}'}
E       assert {'body': '{"a": "123"}'} == {'body': '{"a":"123"}'}
E         
E         Differing items:
E         {'body': '{"a": "123"}'} != {'body': '{"a":"123"}'}
E         Use -v to get more diff

tests/test_requests.py:94: AssertionError
_ test_wsgi_post[trio] _
_ test_request_body[asyncio] _
_ test_request_stream[asyncio] _
_ test_request_body[trio] _
_ test_request_stream[trio] _
= 6 failed, 861 passed, 2 xfailed in 4.02s =
```

It’s easy enough to work around this downstream in Fedora by reverting just https://github.com/encode/starlette/pull/2773/commits/24de2bfc8aa99a084a9b4fcfab1e52d7a6747cd9 in a patch until our `python-httpx` is updated to 0.28.x. In any case, this is a testing issue, not an issue with the Starlette library itself. Still, I’m curious whether this should be considered a bug.

**Comment by Kludex** [comment, marked as answer]
Thanks for the report 🙏

> It’s easy enough to work around this downstream in Fedora by reverting just https://github.com/encode/starlette/commit/24de2bfc8aa99a084a9b4fcfab1e52d7a6747cd9 in a patch until our python-httpx is updated to 0.28.x. In any case, this is a testing issue, not an issue with the Starlette library itself. Still, I’m curious whether this should be considered a bug.

It affects the `TestClient`, which is a public API, and you could say it's a breaking change... Since HTTPX was fine with this change, I think it's all good.

Ref.: 
- https://github.com/encode/httpx/pull/3448
- https://github.com/encode/httpx/pull/3367
